File: webapp/app/routers/benchmark.py
"""Benchmark-API (Task D) — öffentliche Read-Routen + Admin-POST.

Öffentlich (kein Login):
    GET /api/benchmark/meta            — Methodik, Kategorien, Version, Stand
    GET /api/benchmark/samples         — aktive, nicht-held-out Samples
    GET /api/benchmark/audio/{id}      — finale WAV (unkomprimiert, Range)
    GET /api/benchmark/preview/{id}    — Preview-MP3 (128k, on-demand)
    GET /api/benchmark/results         — gepoolte Ergebnisse (latest.json)
    GET /api/benchmark/pricing         — Preisvergleich (WER/€)
    GET /api/benchmark/versions        — Versions-History

Admin (require_admin):
    POST /api/benchmark/samples/{id}/reject  — Ablehnen → Auto-Ersatz + vN+1
    POST /api/benchmark/samples/{id}/edit    — Metadaten editieren
"""
from __future__ import annotations

import hashlib
import hmac
from pathlib import Path
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import FileResponse, JSONResponse, Response
from pydantic import BaseModel

from ..benchmark_service import BenchmarkService
from ..config import settings
from ..deps import require_admin

router = APIRouter(prefix="/api/benchmark", tags=["benchmark"])


def _service() -> BenchmarkService:
    return BenchmarkService(settings.BENCHMARK_DATA_DIR)


def _require_data() -> BenchmarkService:
    """404 statt Crash, wenn noch keine Benchmark-Daten existieren."""
    svc = _service()
    try:
        svc.latest_manifest()
    except FileNotFoundError:
        raise HTTPException(404, "no benchmark data available")
    return svc


# ── Shared-Key-Auth (Change 031) ─────────────────────────────────────────

def _benchmark_keys() -> List[str]:
    """Konfigurierte Keys (kommasepariert, getrimmt). Leer = deaktiviert."""
    return [k.strip() for k in settings.BENCHMARK_API_KEYS.split(",") if k.strip()]


def _authenticated_key(request: Request) -> Optional[str]:
    """Gibt den authentifizierten Key zurück oder None (401-würdig)."""
    keys = _benchmark_keys()
    if not keys:
        raise HTTPException(503, "benchmark api not configured")
    auth = request.headers.get("Authorization", "")
    token = auth[len("Bearer "):].strip() if auth.lower().startswith("bearer ") else ""
    if not token:
        return None
    for k in keys:
        if hmac.compare_digest(token, k):
            return k
    return None


def require_benchmark_key(request: Request) -> None:
    """Dependency: Shared-Key für Backend↔Webapp-Endpunkte (package/submit)."""
    if _authenticated_key(request) is None:
        raise HTTPException(401, "invalid benchmark key")


async def _verify_submit_signature(request: Request, key: str) -> None:
    """HMAC-SHA256 über den rohen Body (X-Benchmark-Signature, hex)."""
    sig = request.headers.get("X-Benchmark-Signature", "").strip()
    if not sig:
        raise HTTPException(401, "missing signature")
    raw = await request.body()  # Starlette cached den Body (auch nach Parsing)
    expected = hmac.new(key.encode(), raw, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(sig, expected):
        raise HTTPException(401, "invalid signature")


# ── Öffentliche GET-Routen ────────────────────────────────────────────────


@router.get("/meta")
def meta() -> Dict[str, Any]:
    svc = _require_data()
    m = svc.latest_manifest()
    public = svc.public_samples()
    # 2-Achsen-Matrix: {kanal: {inhalt: count}} über öffentliche Samples
    matrix: Dict[str, Dict[str, int]] = {}
    for s in public:
        kanal = s.get("kanal") or "clean"
        inhalt = s.get("inhalt") or "allgemein"
        matrix.setdefault(kanal, {}).setdefault(inhalt, 0)
        matrix[kanal][inhalt] += 1
    return {
        "version": m["version"],
        "created_at": m.get("created_at"),
        "supersedes": m.get("supersedes"),
        "categories": m.get("categories", []),
        "sample_count": len(public),
        "per_category": _count_by_category(public),
        "axes": m.get("axes"),
        "matrix": matrix,
        "matrix_total": sum(v for cell in matrix.values() for v in cell.values()),
        "methodology": m.get("methodology", "WER/CER/RTF auf CommonVoice-de + TTS"),
        "disclaimer": m.get(
            "disclaimer",
            "Held-out-Samples und Referenztexte sind nicht öffentlich "
            "(Anti-Gaming). Ergebnisse sind Momentaufnahmen.",
        ),
    }


@router.get("/samples")
def samples() -> Dict[str, Any]:
    svc = _require_data()
    m = svc.latest_manifest()
    public = svc.public_samples()
    return {
        "version": m["version"],
        "samples": [
            {
                "id": s["id"],
                "category": s["category"],
                "text": s["text"],  # Referenztext: öffentlich sichtbar (nicht held-out)
                "accent": s.get("accent", ""),
                "age": s.get("age", ""),
                "preview_url": f"/api/benchmark/preview/{s['id']}",
                "audio_url": f"/api/benchmark/audio/{s['id']}",
            }
            for s in public
        ],
    }


@router.get("/audio/{sample_id}")
def audio(sample_id: str) -> FileResponse:
    svc = _require_data()
    try:
        path = svc.sample_audio_path(sample_id, kind="final")
    except KeyError:
        raise HTTPException(404, "sample not found")
    if not path.exists():
        raise HTTPException(404, "audio not available")
    return FileResponse(path, media_type="audio/wav", filename=f"{sample_id}.wav")


@router.get("/preview/{sample_id}")
def preview(sample_id: str) -> FileResponse:
    svc = _require_data()
    try:
        path = svc.ensure_preview(sample_id)
    except KeyError:
        raise HTTPException(404, "sample not found")
    except FileNotFoundError:
        raise HTTPException(404, "preview not available")
    return FileResponse(path, media_type="audio/mpeg", filename=f"{sample_id}.mp3")


@router.get("/results")
def results() -> Dict[str, Any]:
    """Gepoolte Benchmark-Ergebnisse (results/latest.json) inkl.
    per_category (Change 032 — bei alten Dateien on-the-fly nachgerüstet)."""
    svc = _service()
    try:
        return svc.latest_results()
    except FileNotFoundError:
        raise HTTPException(404, "no benchmark results yet")


@router.get("/pricing")
def pricing() -> Dict[str, Any]:
    """Preisvergleich: Selbstkosten vs. SaaS vs. kommerziell (aus pricing.json + results)."""
    svc = _service()
    p = svc.data_dir / "pricing.json"
    if not p.exists():
        raise HTTPException(404, "no pricing data yet")
    import json
    return json.loads(p.read_text(encoding="utf-8"))


@router.get("/versions")
def versions() -> Dict[str, Any]:
    svc = _service()
    nums = svc.version_numbers()
    out = []
    for n in nums:
        m = svc._load_manifest(n)
        out.append(
            {
                "version": n,
                "created_at": m.get("created_at"),
                "supersedes": m.get("supersedes"),
                "samples": len(m["samples"]),
                "active": sum(1 for s in m["samples"] if s["status"] == "active"),
                "rejected": sum(1 for s in m["samples"] if s["status"] == "rejected"),
            }
        )
    return {"versions": out}


# ── Selbstbedienung (Change 030): Paket + Hash + Submit ───────────────────


@router.get("/package", dependencies=[Depends(require_benchmark_key)])
def package() -> Response:
    """Tarball der aktuellen Benchmark-Version (manifest + audio + preview)
    mit SHA-256 im Header ``X-Benchmark-SHA256`` (``v<N>:<hex>``)."""
    svc = _require_data()
    m = svc.latest_manifest()
    version = m["version"]
    try:
        sha = svc.package_sha256(version)
        data = svc.build_package_tarball(version)
    except FileNotFoundError:
        raise HTTPException(404, "benchmark package not available")
    return Response(
        content=data,
        media_type="application/gzip",
        headers={
            "X-Benchmark-SHA256": f"v{version}:{sha}",
            "Content-Disposition": f'attachment; filename="benchmark-v{version}.tar.gz"',
        },
    )


@router.get("/package/sha256", dependencies=[Depends(require_benchmark_key)])
def package_sha256() -> Dict[str, Any]:
    """Leichtgewichtiger Paket-Hash (Vorab-Prüfung durch Backends)."""
    svc = _require_data()
    m = svc.latest_manifest()
    version = m["version"]
    try:
        sha = svc.package_sha256(version)
    except FileNotFoundError:
        raise HTTPException(404, "benchmark package not available")
    return {"version": version, "manifest_version": version, "sha256": sha}


class SampleResultRow(BaseModel):
    sample_id: str
    hyp: Optional[str] = None
    wer: Optional[float] = None
    cer: Optional[float] = None
    coverage_pct: Optional[float] = None
    rtf: Optional[float] = None


class BenchmarkSubmit(BaseModel):
    backend: str
    settings: str = "auto"
    manifest_version: int
    manifest_sha256: str
    run_id: Optional[str] = None
    generated_at: Optional[str] = None
    rows: List[SampleResultRow]
    meta: Optional[Dict[str, Any]] = None


@router.post("/submit")
async def submit(request: Request, body: BenchmarkSubmit) -> Any:
    """Backend meldet Benchmark-Ergebnisse (Shared-Key + Body-Signatur, Change 031).

    Auth: Authorization: Bearer <key> (BENCHMARK_API_KEYS) plus
    X-Benchmark-Signature: HMAC-SHA256(key, roher Body). Validierung wie 030:
    manifest_version + manifest_sha256 (409), Backend-Name (422).
    """
    key = _authenticated_key(request)
    if key is None:
        raise HTTPException(401, "invalid benchmark key")
    await _verify_submit_signature(request, key)
    svc = _require_data()
    payload = body.model_dump()
    payload["rows"] = [r.model_dump() for r in body.rows]
    res = svc.apply_submission(payload)
    if not res["ok"]:
        if res["reason"] == "manifest mismatch":
            return JSONResponse(status_code=409, content=res)
        if res["reason"] == "unknown backend":
            return JSONResponse(status_code=422, content=res)
        return JSONResponse(status_code=400, content=res)
    return res


# ── Admin-POST-Routen ─────────────────────────────────────────────────────


class SampleEdit(BaseModel):
    text: Optional[str] = None
    category: Optional[str] = None
    accent: Optional[str] = None
    age: Optional[str] = None
    held_out: Optional[bool] = None


@router.post("/samples/{sample_id}/reject", dependencies=[Depends(require_admin)])
def reject_sample(sample_id: str) -> Dict[str, Any]:
    """Lehnt ein Sample ab: Auto-Ersatz aus dem CV-Pool + neue Version vN+1."""
    svc = _require_data()
    m = svc.latest_manifest()
    sample = next((s for s in m["samples"] if s["id"] == sample_id), None)
    if sample is None:
        raise HTTPException(404, "sample not found")

    used_paths = {s["source_path"] for s in m["samples"]}
    used_ids = {s["id"] for s in m["samples"]}
    replacement = svc.replace_rejected_sample(
        category=sample["category"],
        exclude_ids=used_ids,
        used_paths=used_paths,
        seed=42,
    )
    if replacement is None:
        raise HTTPException(409, "kein Ersatz-Sample im CV-Pool verfügbar")

    new_manifest = svc.create_version_after_reject(sample_id, replacement)
    return {
        "ok": True,
        "old_version": m["version"],
        "new_version": new_manifest["version"],
        "rejected": sample_id,
        "replacement": replacement["id"],
        "replacement_source": replacement["source_path"],
    }


@router.post("/samples/{sample_id}/edit", dependencies=[Depends(require_admin)])
def edit_sample(sample_id: str, body: SampleEdit) -> Dict[str, Any]:
    """Editiert Metadaten der aktuellen Version in-place."""
    svc = _require_data()
    fields = {k: v for k, v in body.model_dump().items() if v is not None}
    if not fields:
        raise HTTPException(400, "keine Felder zum Editieren")
    try:
        s = svc.edit_sample(sample_id, **fields)
    except KeyError:
        raise HTTPException(404, "sample not found")
    except ValueError as e:
        raise HTTPException(400, str(e))
    return {"ok": True, "sample": s}


# ── Helfer ────────────────────────────────────────────────────────────────


def _count_by_category(samples: List[dict]) -> Dict[str, int]:
    out: Dict[str, int] = {}
    for s in samples:
        out[s["category"]] = out.get(s["category"], 0) + 1
    return out

File: webapp/frontend/src/benchmark.ts
// Benchmark-API-Wrapper + Pfad-Erkennung für die öffentliche /benchmark-Seite.

export interface BenchmarkCategory {
  id: string;
  name: string;
  description?: string;
}

export interface BenchmarkSample {
  id: string;
  category: string;
  text: string;
  accent?: string;
  age?: string;
  kanal?: string;
  inhalt?: string;
  quelle?: string;
  preview_url: string;
  audio_url: string;
}

export interface BenchmarkAxis {
  beschreibung: string;
  kategorien: Record<string, { name: string }>;
}

export interface BenchmarkMeta {
  version: number;
  created_at?: string;
  supersedes?: number | null;
  categories: BenchmarkCategory[];
  sample_count: number;
  per_category: Record<string, number>;
  axes?: { kanal: BenchmarkAxis; inhalt: BenchmarkAxis };
  matrix?: Record<string, Record<string, number>>;
  matrix_total?: number;
  methodology?: string;
  disclaimer?: string;
}

export interface BenchmarkSamplesResponse {
  version: number;
  samples: BenchmarkSample[];
}

export interface PriceRow {
  backend: string;
  group: "polyschnack" | "commercial";
  wer?: number | null;
  eur_per_min_selfhost?: number | null;
  eur_per_min_saas?: number | null;
  eur_per_min_commercial?: number | null;
}

export interface BenchmarkPricing {
  generated_at?: string;
  rows: PriceRow[];
}

export interface ResultRow {
  backend: string;
  settings?: string;
  sample_id?: string;
  category?: string;
  wer?: number | null;
  cer?: number | null;
  coverage_pct?: number | null;
  rtf?: number | null;
  eur_per_min?: number | null;
}

/** REQ-BEN-046: gepoolte Qualität je (Kategorie × Backend) aus latest.json. */
export interface BenchmarkPerCategoryRow {
  category: string;
  backend: string;
  wer: number;
  cer: number;
  n: number;
}

/** Change 062/065: gepoolte VAD-Ergebnisse (kind="vad" aus latest.json). */
export interface VadResultRow {
  backend: string;
  kind: "vad";
  /** Change 065: Testset-Version (z. B. "v4-public") + Release-Link. */
  testset_version?: string;
  testset_release_url?: string;
  n_samples: number;
  vad_f1_mean: number;
  boundary_start_ms_median: number;
  boundary_end_ms_median: number;
  fp_time_s: number;
  rtf_mean: number;
}

export interface BenchmarkResults {
  version?: number;
  run_id?: string;
  generated_at?: string;
  rows: ResultRow[];
  per_category?: BenchmarkPerCategoryRow[];
  /** Change 039: sample_id -> { backend: wer } für Sample-Mini-Tabellen. */
  per_sample?: Record<string, Record<string, number>>;
  /** Change 062: VAD-Modell-Ergebnisse (getrennt vom ASR-Pool). */
  vad?: VadResultRow[];
}

/** Change 073: VAD-Testset-Sample (öffentlich anhörbar auf der Benchmark-Seite). */
export interface VadSample {
  id: string;
  source: string;
  variant: string;
  split: string;
  /** true = Sample hat Ground-Truth-Sprachregionen (GT); FP-Samples nicht. */
  has_gt: boolean;
  preview_url: string;
  audio_url: string;
}

export interface VadSamplesResponse {
  samples: VadSample[];
  count: number;
}

/** Change 075: Status des Benchmark-Set-Auto-Updates (öffentlich, keine Secrets). */
export interface AvailableSet {
  version: number;
  tag: string;
  published_at?: string | null;
  zip_url?: string;
  zip_size?: number | null;
  sha_url?: string | null;
}

export interface BenchmarkSetStatus {
  mechanism: string;
  configured: boolean;
  /** Change 076: true = env-URL-Pinning aktiv (kein Discovery). */
  pinning_mode: boolean;
  repo: string;
  url: string;
  /** SHA256 nur als 8-Zeichen-Präfix (kein voller Hash nach außen). */
  sha_prefix: string;
  auto_install: boolean;
  current_version: number | null;
  installed_versions: number[];
  /** Change 076: verfügbare Releases aus dem Repo (gecacht). */
  available: AvailableSet[];
  last_error: string | null;
}

export interface SetInstallResponse {
  ok: boolean;
  skipped?: boolean;
  reason?: string;
  installed_version?: number | null;
  current_version?: number | null;
  sha256?: string;
  sample_count?: number;
  supersedes?: number | null;
}

export function parseBenchmarkPath(path: string): boolean {
  const p = path.split("?")[0].replace(/\/+$/, "");
  return p === "/benchmark" || p.startsWith("/benchmark/");
}

async function checkOk(res: Response): Promise<Response> {
  if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
  return res;
}

export async function fetchBenchmarkMeta(): Promise<BenchmarkMeta> {
  return fetch("/api/benchmark/meta").then(checkOk).then((r) => r.json());
}

export async function fetchBenchmarkSamples(): Promise<BenchmarkSamplesResponse> {
  return fetch("/api/benchmark/samples").then(checkOk).then((r) => r.json());
}

/** Change 073: VAD-Testset-Samples (öffentliche Liste mit Audio-URLs). */
export async function fetchVadSamples(): Promise<VadSamplesResponse | null> {
  const res = await fetch("/api/benchmark/vadsamples");
  if (res.status === 404) return null;
  return checkOk(res).then((r) => r.json());
}

export async function fetchBenchmarkResults(): Promise<BenchmarkResults | null> {
  const res = await fetch("/api/benchmark/results");
  if (res.status === 404) return null;
  return checkOk(res).then((r) => r.json());
}

export async function fetchBenchmarkPricing(): Promise<BenchmarkPricing | null> {
  const res = await fetch("/api/benchmark/pricing");
  if (res.status === 404) return null;
  return checkOk(res).then((r) => r.json());
}

export async function fetchBenchmarkVersions(): Promise<{ versions: Array<{ version: number; created_at?: string; active: number; rejected: number }> } | null> {
  const res = await fetch("/api/benchmark/versions");
  if (res.status === 404) return null;
  return checkOk(res).then((r) => r.json());
}

export async function rejectBenchmarkSample(sampleId: string): Promise<{ new_version: number; replacement: string }> {
  return fetch(`/api/benchmark/samples/${encodeURIComponent(sampleId)}/reject`, {
    method: "POST",
  }).then(checkOk).then((r) => r.json());
}

export async function editBenchmarkSample(
  sampleId: string,
  fields: { text?: string; category?: string; accent?: string; age?: string; held_out?: boolean },
): Promise<{ ok: boolean; sample: BenchmarkSample }> {
  return fetch(`/api/benchmark/samples/${encodeURIComponent(sampleId)}/edit`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(fields),
  }).then(checkOk).then((r) => r.json());
}

/** Change 075: Status des Set-Updaters (öffentlich). */
export async function fetchBenchmarkSetStatus(): Promise<BenchmarkSetStatus | null> {
  const res = await fetch("/api/benchmark/sets");
  if (res.status === 404) return null;
  return checkOk(res).then((r) => r.json());
}

/** Change 075: Benchmark-Set installieren (Admin). */
export async function installBenchmarkSet(
  url?: string,
  sha256?: string,
  repo?: string,
  version?: number,
): Promise<SetInstallResponse> {
  return fetch("/api/benchmark/sets/install", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      url: url ?? undefined,
      sha256: sha256 ?? undefined,
      repo: repo ?? undefined,
      version: version ?? undefined,
    }),
  }).then(checkOk).then((r) => r.json());
}
